pandas/_libs/missing.pyx

```diff
@@ -290,12 +291,25 @@ def _create_binary_propagating_op(name, divmod=False):
 
     def method(self, other):
         if (other is C_NA or isinstance(other, str)
-                or isinstance(other, (numbers.Number, np.bool_))):
+                or isinstance(other, (numbers.Number, np.bool_))
```

Comment:
util.is_bool_object for the second one of these.  Is numbers.Number something like Union[float, int, complex] that we could use the optimizied funcs for?

util.is_array for the isinstance(other, np.ndarray) check.  Even though it is syntactically OK without extra parens, can you put a pair around the isinstance+shape check

Reply:
Numbers is appropriate since it’s an ABC others can register on.

Reply:
I personally prefer the explicit `isinstance` checks here. It is already complex to fully oversee what types are handled, and using those helper methods only makes it more opaque.

Reply:
in general i agree, but in the cython code we generally lean towards optimizing harder.  no strong opinion.

Reply:
That's also a good point (cython does not optimize those automatically like that I suppose?). But note that this is not the most performance sensitive code (it's not in the inner loop of a vectorized operation or so)